### PR TITLE
add kubemacpool to the operator

### PR DESCRIPTION
**Pull request by SchSeba**
This PR add the kubemacpool into the operator.

---
This change is [Reviewable](https://reviewable.io/reviews/kubevirt/cluster-network-addons-operator/8)

### CODE REVIEW

**Comment by SchSeba** [comment]
This PR is not ready for review yet there are some open issues in the kubemacpool project to close first

File: config-example.yaml [hunk 1]
```diff
@@ -13,3 +13,6 @@ spec:
         }
       }]
   linuxBridge: {}
+  kubeMacPool:
+   endPoolRange: "FD:FF:FF:FF:FF:FF"
```
Comment (phoracek) [inline config-example.yaml]: please switch those two lines, first start, then end :)

File: data/kubemacpool/002-configmap.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+---
+apiVersion: v1
+{{ if .DataExist }}
+data:
+  END_POOL_RANGE: {{ .EndPoolRange }}
+  START_POOL_RANGE: {{ .StartPoolRange}}
+{{ end}}
```
Comment (phoracek) [inline data/kubemacpool/002-configmap.yaml]: {{ end }}

File: data/kubemacpool/002-configmap.yaml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+---
+apiVersion: v1
+{{ if .DataExist }}
+data:
+  END_POOL_RANGE: {{ .EndPoolRange }}
+  START_POOL_RANGE: {{ .StartPoolRange}}
```
Comment (phoracek) [inline data/kubemacpool/002-configmap.yaml]: {{ .StartPoolRange }} and switch those two

File: data/kubemacpool/004-statefulset.yaml [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+              valueFrom:
+                configMapKeyRef:
+                  key: START_POOL_RANGE
+                  name: kubemacpool-mac-range-config
+            - name: END_POOL_RANGE
+              valueFrom:
+                configMapKeyRef:
+                  key: END_POOL_RANGE
+                  name: kubemacpool-mac-range-config
+          image: {{.KubeMacPoolImage}}
```
Comment (phoracek) [inline data/kubemacpool/004-statefulset.yaml]: {{ .KubeMacPoolImage }}

File: data/kubemacpool/004-statefulset.yaml [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+          imagePullPolicy: {{.ImagePullPolicy}}
```
Comment (phoracek) [inline data/kubemacpool/004-statefulset.yaml]: {{ .ImagePullPolicy }}

File: deploy/cluster-network-addons-operator_03_deployment.yaml [hunk 6]
```diff
@@ -30,7 +30,9 @@ spec:
         - name: MULTUS_IMAGE
           value: "docker.io/nfvpe/multus:latest"
         - name: LINUX_BRIDGE_IMAGE
-          value: "quay.io/kubevirt/cni-default-plugins"
+          value: "quay.io/kubevirt/cni-default-plugins:latest"
```
Comment (phoracek) [inline deploy/cluster-network-addons-operator_03_deployment.yaml]: should be in a separate PR

File: pkg/apis/networkaddonsoperator/v1alpha1/networkaddonsconfig_types.go [hunk 7]
```diff
@@ -18,6 +19,11 @@ type Multus struct {
 
 type LinuxBridge struct{}
 
+type KubeMacPool struct {
+	EndPoolRange   string `json:"endPoolRange,omitempty"`
+	StartPoolRange string `json:"startPoolRange,omitempty"`
```
Comment (phoracek) [inline pkg/apis/networkaddonsoperator/v1alpha1/networkaddonsconfig_types.go]: switch those two lines

File: data/kubemacpool/002-configmap.yaml [hunk 8]
```diff
@@ -0,0 +1,14 @@
+---
+apiVersion: v1
+{{ if .DataExist }}
+data:
+  END_POOL_RANGE: {{ .EndPoolRange }}
```
Comment (phoracek) [inline data/kubemacpool/002-configmap.yaml]: what happens if not provided? shouldn't we fall back to a default value?

File: pkg/network/kubemacpool.go [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+		return nil, nil
+	}
+
+	// render the manifests on disk
+	data := render.MakeRenderData()
+	data.Data["KubeMacPoolImage"] = os.Getenv("KUBEMACPOOL_IMAGE")
+	data.Data["ImagePullPolicy"] = os.Getenv("IMAGE_PULL_POLICY")
+
+	if conf.KubeMacPool.StartPoolRange == "" || conf.KubeMacPool.EndPoolRange == "" {
+		data.Data["DataExist"] = false
```
Comment (phoracek) [inline pkg/network/kubemacpool.go]: When I see DataExist in the template, it does not say much, would you please rename it from "condition" to "argument", maybe something like "ConfigurePoolRange".

File: data/kubemacpool/002-configmap.yaml (hunk 8 above)
Comment (SchSeba) [reply data/kubemacpool/002-configmap.yaml]: I want to implement the logic of the default value into the mac pool and not here.

**Comment by SchSeba** [comment]
@phoracek Done. when you have time please take another look but don't merge this yet I want to host the kubemacpool in a proper repo first

**Comment by phoracek** [comment]
@SchSeba please update the README with macpool documentation, other than that it looks good now

**Comment by SchSeba** [comment]
@phoracek I think the PR is ready now.
I also added the requested documentation

File: pkg/network/network.go [hunk 10]
```diff
@@ -57,6 +57,7 @@ func IsChangeSafe(prev, next *opv1alpha1.NetworkAddonsConfigSpec) error {
 	errs = append(errs, changeSafeLinuxBridge(prev, next)...)
 	errs = append(errs, changeSafeSriov(prev, next)...)
 	errs = append(errs, changeSafeImagePullPolicy(prev, next)...)
+	errs = append(errs, changeSafeKubeMacPool(prev, next)...)
```
Comment (phoracek) [inline pkg/network/network.go]: just a detail, but please put kubemacpool above image pull policy

File: pkg/network/kubemacpool.go [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+// renderLinuxBridge generates the manifests of Linux Bridge
+func renderKubeMacPool(conf *opv1alpha1.NetworkAddonsConfigSpec, manifestDir string) ([]*unstructured.Unstructured, error) {
+	if conf.KubeMacPool == nil {
+		return nil, nil
+	}
+
+	// render the manifests on disk
+	data := render.MakeRenderData()
+	data.Data["KubeMacPoolImage"] = os.Getenv("KUBEMACPOOL_IMAGE")
+	data.Data["ImagePullPolicy"] = os.Getenv("IMAGE_PULL_POLICY")
```
Comment (phoracek) [inline pkg/network/kubemacpool.go]: this should be taken from CR, not from env variables anymore

File: pkg/network/kubemacpool.go [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+	if conf.KubeMacPool.StartPoolRange == "" || conf.KubeMacPool.EndPoolRange == "" {
+		data.Data["ConfigurePoolRange"] = false
+	} else {
+		data.Data["StartPoolRange"] = conf.KubeMacPool.StartPoolRange
+		data.Data["EndPoolRange"] = conf.KubeMacPool.EndPoolRange
+		data.Data["ConfigurePoolRange"] = true
+
```
Comment (phoracek) [inline pkg/network/kubemacpool.go]: drop empty line

File: pkg/network/kubemacpool.go [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+		return nil, nil
+	}
+
+	// render the manifests on disk
+	data := render.MakeRenderData()
+	data.Data["KubeMacPoolImage"] = os.Getenv("KUBEMACPOOL_IMAGE")
+	data.Data["ImagePullPolicy"] = os.Getenv("IMAGE_PULL_POLICY")
+
+	if conf.KubeMacPool.StartPoolRange == "" || conf.KubeMacPool.EndPoolRange == "" {
+		data.Data["ConfigurePoolRange"] = false
```
Comment (phoracek) [inline pkg/network/kubemacpool.go]: this suggests, that if user does not pass MAC range, we don't create the configmap. However, it is not documented in the README.

Anyway, is there a good reason to deploy kubemacpool without the range configured? Maybe we should make it simpler and always create it. Either make the range mandatory or add defaulting.

File: data/kubemacpool/003-secret.yaml [hunk 14]
```diff
@@ -0,0 +1,6 @@
+---
+apiVersion: v1
+kind: Secret
```
Comment (phoracek) [inline data/kubemacpool/003-secret.yaml]: I believe Secrets on OpenShift require privileged pods, hence we would need to create an SCC. Did you test this on OpenShift?

Comment (SchSeba) [reply data/kubemacpool/003-secret.yaml]: don't need any special scc.

oc get po
```
kubemacpool-system                  kubemacpool-mac-controller-manager-0               1/1       Running   1          1m        10.128.0.47      node01    <none>
```

oc logs
```
{"level":"info","ts":1554124103.0487776,"logger":"Main","msg":"Setting up client for manager"}
{"level":"info","ts":1554124103.1602395,"logger":"Main","msg":"kubevirt was not found in the cluster start a watching process"}
{"level":"info","ts":1554124103.160369,"logger":"Main","msg":"Setting up manager"}
{"level":"info","ts":1554124103.3492582,"logger":"Main","msg":"Setting up controller"}
{"level":"info","ts":1554124103.3493834,"logger":"kubebuilder.controller","msg":"Starting EventSource","controller":"pod-controller","source":"kind source: /, Kind="}
{"level":"info","ts":1554124121.0742912,"logger":"kubebuilder.webhook","msg":"installing webhook configuration in cluster"}
{"level":"info","ts":1554124121.0744402,"logger":"kubebuilder.controller","msg":"Starting Controller","controller":"pod-controller"}
{"level":"info","ts":1554124122.349403,"logger":"kubebuilder.controller","msg":"Starting workers","controller":"pod-controller","worker count":1}
{"level":"info","ts":1554124122.5875223,"logger":"kubebuilder.webhook","msg":"starting the webhook server."}
{"level":"info","ts":1554124244.168139,"logger":"PoolManager","msg":"mac from pool was allocated to the pod","allocatedMac":"02:00:00:00:00:01","podName":"samplepod","podNamespace":"default"}
```

File: pkg/network/kubemacpool.go (hunk 11 above)
Comment (SchSeba) [reply pkg/network/kubemacpool.go]: done

File: pkg/network/kubemacpool.go (hunk 13 above)
Comment (SchSeba) [reply pkg/network/kubemacpool.go]: for now the kubemacpool doesn't generate so I remove it from this PR  and make it mandatory.
I will open a separate one after the kubemacpool support it.

I removed the omitempty from them

File: pkg/network/kubemacpool.go (hunk 12 above)
Comment (SchSeba) [reply pkg/network/kubemacpool.go]: done

File: pkg/network/network.go (hunk 10 above)
Comment (SchSeba) [reply pkg/network/network.go]: done

File: pkg/apis/networkaddonsoperator/v1alpha1/networkaddonsconfig_types.go [hunk 15]
```diff
@@ -11,6 +11,7 @@ type NetworkAddonsConfigSpec struct {
 	LinuxBridge     *LinuxBridge `json:"linuxBridge,omitempty"`
 	Sriov           *Sriov       `json:"sriov,omitempty"`
 	ImagePullPolicy string       `json:"imagePullPolicy,omitempty"`
+	KubeMacPool     *KubeMacPool `json:"kubeMacPool,omitempty"`
```
Comment (phoracek) [inline pkg/apis/networkaddonsoperator/v1alpha1/networkaddonsconfig_types.go]: i know it's a nit, but please, put it above ImagePullPolicy

File: pkg/network/kubemacpool.go [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+	"path/filepath"
+	"reflect"
+
+	"github.com/kubevirt/cluster-network-addons-operator/pkg/render"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	opv1alpha1 "github.com/kubevirt/cluster-network-addons-operator/pkg/apis/networkaddonsoperator/v1alpha1"
+)
+
```
Comment (phoracek) [inline pkg/network/kubemacpool.go]: please add validating code making sure both ends of the range are specified

File: config-example.yaml [hunk 17]
```diff
@@ -3,7 +3,10 @@ kind: NetworkAddonsConfig
 metadata:
   name: cluster
 spec:
-  multus: {}
-  linuxBridge: {}
-  sriov: {}
+  #multus: {}
```
Comment (phoracek) [inline config-example.yaml]: you forgot this

File: config-example.yaml [hunk 18]
```diff
@@ -3,7 +3,10 @@ kind: NetworkAddonsConfig
...
+  #linuxBridge: {}
+  #sriov: {}
   imagePullPolicy: Always
+  kubeMacPool:
```
Comment (phoracek) [inline config-example.yaml]: please move it above imagePullPolicy

File: pkg/network/kubemacpool.go [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+// ValidateMultus validates the combination of DisableMultiNetwork and AddtionalNetworks
+func validateKubeMacPool(conf *opv1alpha1.NetworkAddonsConfigSpec) []error {
+	if conf.Multus == nil {
```
Comment (phoracek) [inline pkg/network/kubemacpool.go]: bad copy paste

File: pkg/network/kubemacpool.go [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+		return []error{}
+	}
+
+	if conf.KubeMacPool.StartPoolRange == "" {
+		return []error{errors.Errorf("StartPoolRange must be configured")}
```
Comment (phoracek) [inline pkg/network/kubemacpool.go]: **s**tartPoolRange

File: pkg/network/kubemacpool.go [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+	} else if _, err := net.ParseMAC(conf.KubeMacPool.StartPoolRange); err != nil {
+		return []error{errors.Errorf("failed to parse StartPoolRange invalid mac address")}
+	}
+
+	if conf.KubeMacPool.EndPoolRange == "" {
+		return []error{errors.Errorf("EndPoolRange must be configured")}
```
Comment (phoracek) [inline pkg/network/kubemacpool.go]: **e**ndPoolRange

File: pkg/network/kubemacpool.go [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+	} else if _, err := net.ParseMAC(conf.KubeMacPool.EndPoolRange); err != nil {
+		return []error{errors.Errorf("failed to parse EndPoolRange invalid mac address")}
```
Comment (phoracek) [inline pkg/network/kubemacpool.go]: same

File: pkg/network/kubemacpool.go [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+// ValidateMultus validates the combination of DisableMultiNetwork and AddtionalNetworks
+func validateKubeMacPool(conf *opv1alpha1.NetworkAddonsConfigSpec) []error {
+	if conf.Multus == nil {
+		return []error{}
+	}
+
+	if conf.KubeMacPool.StartPoolRange == "" {
+		return []error{errors.Errorf("StartPoolRange must be configured")}
+	} else if _, err := net.ParseMAC(conf.KubeMacPool.StartPoolRange); err != nil {
+		return []error{errors.Errorf("failed to parse StartPoolRange invalid mac address")}
```
Comment (phoracek) [inline pkg/network/kubemacpool.go]: same

File: pkg/network/network.go [hunk 24]
```diff
@@ -23,6 +23,7 @@ func Validate(conf *opv1alpha1.NetworkAddonsConfigSpec, openshiftNetworkConfig *
 
 	errs = append(errs, validateMultus(conf, openshiftNetworkConfig)...)
 	errs = append(errs, validateImagePullPolicy(conf)...)
+	errs = append(errs, validateKubeMacPool(conf)...)
```
Comment (phoracek) [inline pkg/network/network.go]: please move above image pull policy.

i'd like to keep image pull policy on bottom everywhere since it is not a component like all the others, it's just additional lifecycle parameter
